File: 3d/src/main/java/io/github/xerooup/frame3d/graphics/mesh/BoxMesh.java
package io.github.xerooup.frame3d.graphics.mesh;

import io.github.xerooup.frame3d.graphics.Color;

public class BoxMesh extends BaseMesh {
    private final float[] vertices;
    private final int[] indices;

    public BoxMesh() {
        this(Color.WHITE);
    }

    public BoxMesh(Color color) {
        this(1.0f, 1.0f, 1.0f, color);
    }

    public BoxMesh(float width, float height, float depth, Color color) {
        float w = width / 2;
        float h = height / 2;
        float d = depth / 2;
        float r = color.r;
        float g = color.g;
        float b = color.b;

        // Vertices: position (3) + color (3) + normal (3) + texCoord (2) = 11 floats
        vertices = new float[] {
                // Front face (normal: 0, 0, 1)
                -w, -h,  d,  r, g, b,  0, 0, 1,  0, 0,
                w, -h,  d,  r, g, b,  0, 0, 1,  1, 0,
                w,  h,  d,  r, g, b,  0, 0, 1,  1, 1,
                -w,  h,  d,  r, g, b,  0, 0, 1,  0, 1,

                // Back face (normal: 0, 0, -1)
                -w, -h, -d,  r, g, b,  0, 0, -1,  1, 0,
                w, -h, -d,  r, g, b,  0, 0, -1,  0, 0,
                w,  h, -d,  r, g, b,  0, 0, -1,  0, 1,
                -w,  h, -d,  r, g, b,  0, 0, -1,  1, 1,

                // Top face (normal: 0, 1, 0)
                -w,  h, -d,  r, g, b,  0, 1, 0,  0, 0,
                -w,  h,  d,  r, g, b,  0, 1, 0,  0, 1,
                w,  h,  d,  r, g, b,  0, 1, 0,  1, 1,
                w,  h, -d,  r, g, b,  0, 1, 0,  1, 0,

                // Bottom face (normal: 0, -1, 0)
                -w, -h, -d,  r, g, b,  0, -1, 0,  0, 1,
                w, -h, -d,  r, g, b,  0, -1, 0,  1, 1,
                w, -h,  d,  r, g, b,  0, -1, 0,  1, 0,
                -w, -h,  d,  r, g, b,  0, -1, 0,  0, 0,

                // Right face (normal: 1, 0, 0)
                w, -h, -d,  r, g, b,  1, 0, 0,  0, 0,
                w,  h, -d,  r, g, b,  1, 0, 0,  0, 1,
                w,  h,  d,  r, g, b,  1, 0, 0,  1, 1,
                w, -h,  d,  r, g, b,  1, 0, 0,  1, 0,

                // Left face (normal: -1, 0, 0)
                -w, -h, -d,  r, g, b,  -1, 0, 0,  1, 0,
                -w, -h,  d,  r, g, b,  -1, 0, 0,  0, 0,
                -w,  h,  d,  r, g, b,  -1, 0, 0,  0, 1,
                -w,  h, -d,  r, g, b,  -1, 0, 0,  1, 1
        };

        indices = new int[] {
                0,  1,  2,  2,  3,  0,   // Front
                4,  5,  6,  6,  7,  4,   // Back
                8,  9, 10, 10, 11,  8,   // Top
                12, 13, 14, 14, 15, 12,  // Bottom
                16, 17, 18, 18, 19, 16,  // Right
                20, 21, 22, 22, 23, 20   // Left
        };

        setup(vertices, indices);
    }

    @Override
    public float[] getVertices() {
        return vertices;
    }

    @Override
    public int[] getIndices() {
        return indices;
    }
}

File: 3d/src/main/java/io/github/xerooup/frame3d/graphics/mesh/BoxMesh.kt
package io.github.xerooup.frame3d.graphics.mesh

import io.github.xerooup.frame3d.graphics.Color

class BoxMesh(color: Color = Color.WHITE) : BaseMesh() {
    constructor() : this(Color.WHITE)
    constructor(width: Float, height: Float, depth: Float, color: Color) : this(color) {
        // Recreate with custom size
        val w = width / 2
        val h = height / 2
        val d = depth / 2
        val r = color.r
        val g = color.g
        val b = color.b

        val customVertices = floatArrayOf(
            // Front face
            -w, -h,  d,  r, g, b,  0f, 0f, 1f,  0f, 0f,
            w, -h,  d,  r, g, b,  0f, 0f, 1f,  1f, 0f,
            w,  h,  d,  r, g, b,  0f, 0f, 1f,  1f, 1f,
            -w,  h,  d,  r, g, b,  0f, 0f, 1f,  0f, 1f,

            // Back face
            -w, -h, -d,  r, g, b,  0f, 0f, -1f,  1f, 0f,
            w, -h, -d,  r, g, b,  0f, 0f, -1f,  0f, 0f,
            w,  h, -d,  r, g, b,  0f, 0f, -1f,  0f, 1f,
            -w,  h, -d,  r, g, b,  0f, 0f, -1f,  1f, 1f,

            // Top face
            -w,  h, -d,  r, g, b,  0f, 1f, 0f,  0f, 0f,
            -w,  h,  d,  r, g, b,  0f, 1f, 0f,  0f, 1f,
            w,  h,  d,  r, g, b,  0f, 1f, 0f,  1f, 1f,
            w,  h, -d,  r, g, b,  0f, 1f, 0f,  1f, 0f,

            // Bottom face
            -w, -h, -d,  r, g, b,  0f, -1f, 0f,  0f, 1f,
            w, -h, -d,  r, g, b,  0f, -1f, 0f,  1f, 1f,
            w, -h,  d,  r, g, b,  0f, -1f, 0f,  1f, 0f,
            -w, -h,  d,  r, g, b,  0f, -1f, 0f,  0f, 0f,

            // Right face
            w, -h, -d,  r, g, b,  1f, 0f, 0f,  0f, 0f,
            w,  h, -d,  r, g, b,  1f, 0f, 0f,  0f, 1f,
            w,  h,  d,  r, g, b,  1f, 0f, 0f,  1f, 1f,
            w, -h,  d,  r, g, b,  1f, 0f, 0f,  1f, 0f,

            // Left face
            -w, -h, -d,  r, g, b,  -1f, 0f, 0f,  1f, 0f,
            -w, -h,  d,  r, g, b,  -1f, 0f, 0f,  0f, 0f,
            -w,  h,  d,  r, g, b,  -1f, 0f, 0f,  0f, 1f,
            -w,  h, -d,  r, g, b,  -1f, 0f, 0f,  1f, 1f
        )

        setup(customVertices, indices)
    }

    private val w = 0.5f
    private val h = 0.5f
    private val d = 0.5f
    private val r = color.r
    private val g = color.g
    private val b = color.b

    private val vertices = floatArrayOf(
        // Front face
        -w, -h,  d,  r, g, b,  0f, 0f, 1f,  0f, 0f,
        w, -h,  d,  r, g, b,  0f, 0f, 1f,  1f, 0f,
        w,  h,  d,  r, g, b,  0f, 0f, 1f,  1f, 1f,
        -w,  h,  d,  r, g, b,  0f, 0f, 1f,  0f, 1f,

        // Back face
        -w, -h, -d,  r, g, b,  0f, 0f, -1f,  1f, 0f,
        w, -h, -d,  r, g, b,  0f, 0f, -1f,  0f, 0f,
        w,  h, -d,  r, g, b,  0f, 0f, -1f,  0f, 1f,
        -w,  h, -d,  r, g, b,  0f, 0f, -1f,  1f, 1f,

        // Top face
        -w,  h, -d,  r, g, b,  0f, 1f, 0f,  0f, 0f,
        -w,  h,  d,  r, g, b,  0f, 1f, 0f,  0f, 1f,
        w,  h,  d,  r, g, b,  0f, 1f, 0f,  1f, 1f,
        w,  h, -d,  r, g, b,  0f, 1f, 0f,  1f, 0f,

        // Bottom face
        -w, -h, -d,  r, g, b,  0f, -1f, 0f,  0f, 1f,
        w, -h, -d,  r, g, b,  0f, -1f, 0f,  1f, 1f,
        w, -h,  d,  r, g, b,  0f, -1f, 0f,  1f, 0f,
        -w, -h,  d,  r, g, b,  0f, -1f, 0f,  0f, 0f,

        // Right face
        w, -h, -d,  r, g, b,  1f, 0f, 0f,  0f, 0f,
        w,  h, -d,  r, g, b,  1f, 0f, 0f,  0f, 1f,
        w,  h,  d,  r, g, b,  1f, 0f, 0f,  1f, 1f,
        w, -h,  d,  r, g, b,  1f, 0f, 0f,  1f, 0f,

        // Left face
        -w, -h, -d,  r, g, b,  -1f, 0f, 0f,  1f, 0f,
        -w, -h,  d,  r, g, b,  -1f, 0f, 0f,  0f, 0f,
        -w,  h,  d,  r, g, b,  -1f, 0f, 0f,  0f, 1f,
        -w,  h, -d,  r, g, b,  -1f, 0f, 0f,  1f, 1f
    )

    private val indices = intArrayOf(
        0,  1,  2,  2,  3,  0,   // Front
        4,  5,  6,  6,  7,  4,   // Back
        8,  9, 10, 10, 11,  8,   // Top
        12, 13, 14, 14, 15, 12,  // Bottom
        16, 17, 18, 18, 19, 16,  // Right
        20, 21, 22, 22, 23, 20   // Left
    )

    init {
        setup(vertices, indices)
    }

    override fun getVertices() = vertices
    override fun getIndices() = indices
}